pkg/messaging/nats/pubsub.go

```diff
@@ -77,7 +78,16 @@ func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler)
 	case true:
 		// Check topic ID
 		if _, ok := s[id]; ok {
-			return ErrAlreadySubscribed
+			// Unlocking, so that Unsubscribe() can access ps.subscriptions
+			ps.mu.Unlock()
+			if err := ps.Unsubscribe(id, topic); err != nil {
+				return err
+			}
+			ps.mu.Lock()
```

Comment:
You need to be **very** careful with locks. Where do you unlock this one, for example (there is no more `defer` at the beginning to save you)?

Reply:
In this case, I can add:
```
if err != nil {
  ps.mu.Lock()
  return err
```

I was assuming unsubscribe shoudn't fail, since if the subscription already exists, that means the values are correct, and unsubscribe will have no problem.
That was wrong of me to assume, I'll fix this ASAP.

For the ps.mu.Lock() below that, it's unlocked by the defer unlock in the beginning.
Earlier I was using multiple lock and unlock statements inside the switch blocks. But Drasko suggested that it looked untidy, and unreadable. So, we switched to this cleaner approach.
The tests were all passing, and I even tried it with my ws-adapter code.

Reply:
`I was assuming unsubscribe shouldn't fail, since if the subscription already exists, that means the values are correct, and unsubscribe will have no problem.` Exactly, that's not something you can control. You can solve it whichever way you want. Hint:  you can call `defer`  from anywhere, it does not have to be at the begging of the function (while that's the usual pattern).